File: apps/android/app/src/main/kotlin/ai/parkshield/android/feature/auth/data/SecureSessionStore.kt
package ai.parkshield.android.feature.auth.data

import ai.parkshield.android.core.security.SecureStorageException
import ai.parkshield.android.core.security.SecureValueStore
import ai.parkshield.android.feature.auth.domain.AuthFailure
import ai.parkshield.android.feature.auth.domain.AuthResult
import ai.parkshield.android.feature.auth.domain.TokenPair

class SecureSessionStore(private val store: SecureValueStore) {
    suspend fun save(pair: TokenPair): AuthResult<Unit> = try {
        // Write refresh first; access is only useful with a matching durable refresh token.
        store.write(REFRESH, pair.refreshToken)
        store.write(ACCESS, pair.accessToken)
        AuthResult.Success(Unit)
    } catch (_: SecureStorageException) { clear(); AuthResult.Failure(AuthFailure.SecureStorage) }

    suspend fun refreshToken(): AuthResult<String?> = try { AuthResult.Success(store.read(REFRESH)) }
    catch (_: SecureStorageException) { AuthResult.Failure(AuthFailure.SecureStorage) }

    suspend fun accessToken(): AuthResult<String?> = try { AuthResult.Success(store.read(ACCESS)) }
    catch (_: SecureStorageException) { AuthResult.Failure(AuthFailure.SecureStorage) }

    suspend fun clear() { runCatching { store.remove(ACCESS) }; runCatching { store.remove(REFRESH) } }

    private companion object { const val ACCESS = "access-token-v1"; const val REFRESH = "refresh-token-v1" }
}

File: apps/ios/Sources/ParkShieldFoundation/Feature/Auth/Data/SecureSessionStore.swift
import Foundation

public struct SecureSessionStore: Sendable {
    private let store: any SecureValueStore
    public init(store: any SecureValueStore) { self.store = store }
    func save(_ pair: TokenPair) async -> Result<Void, AuthFailure> { do { try await store.write(pair.refreshToken, for: "refresh-token-v1"); try await store.write(pair.accessToken, for: "access-token-v1"); return .success(()) } catch { await clear(); return .failure(.secureStorage) } }
    func refreshToken() async -> Result<String?, AuthFailure> { do { return .success(try await store.read(for: "refresh-token-v1")) } catch { return .failure(.secureStorage) } }
    func accessToken() async -> Result<String?, AuthFailure> { do { return .success(try await store.read(for: "access-token-v1")) } catch { return .failure(.secureStorage) } }
    func clear() async { try? await store.removeValue(for: "access-token-v1"); try? await store.removeValue(for: "refresh-token-v1") }
}
